doc/guides: add guide for porting new MCUs (#22173)
RIOT-OS/RIOT: doc/guides/advanced_tutorials/porting_cpus.mdx

---
title: Porting CPUs
description: Guide on how to port new CPU families to RIOT-OS
---

> import { FileTree } from '@astrojs/starlight/components';

Reviewer: ```suggestion
import FileTree from '@components/FileTree.astro';
```

To fix the next occurring issue:

```
11:11:12 [ERROR] Cannot find module '@astrojs/starlight/components' imported from '/home/cbuec/RIOTstuff/riot-vanillaice/RIOT/doc/guides/advanced_tutorials/porting_cpus.mdx'
  Stack trace:
    at fetchModule (file:///home/cbuec/RIOTstuff/riot-vanillaice/RIOT/doc/starlight/node_modules/vite/dist/node/chunks/config.js:33980:34)
    [...] See full stack trace in the browser, or rerun with --verbose.
```

Reviewer: I just saw that I recommended this line. Sorry for suggesting something incorrect 🫨

Reviewer: No, it was @AnnsAnns, not me 👀

The goal of a CPU port is to make the MCU family usable from RIOT's generic
build system, startup code and peripheral APIs. A complete port
usually consists of:

- a `cpu/<cpu>/` directory that provides startup, common CPU configuration, and
  peripheral implementations,
- one or more `boards/<board>/` directories that map the CPU to a concrete
  development board, and
- documentation and tests that make the port maintainable.

This document focuses on the CPU side. For the board side, also refer to the
existing RIOT guide on [porting boards](/advanced_tutorials/porting_boards/).

:::note
A CPU port defines what the MCU family can do. A board port defines which of
those features are actually available on a specific board.
:::

## CPU Port File Structure

A typical CPU port has a file structure similar to the following:

Entries marked with `*` are optional and depend on the CPU family and reuse
strategy. Entries without `*` are typically required for a minimal CPU port.

<FileTree>
- cpu/foo
  - include
    - cpu_conf.h
    - periph_cpu.h
    - vendor*
  - ldscripts*
  - periph*
    - gpio.c
    - i2c.c
    - spi.c
    - …
  - vectors

>     - vectors_<model>.c

Reviewer: ```suggestion
    - vectors_\<model\>.c
```

To fix the next error:

```
11:10:05 [ERROR] Expected a closing tag for `<model>` (50:15-50:22) before the end of `paragraph`
  Stack trace:
    at /home/cbuec/RIOTstuff/riot-vanillaice/RIOT/doc/guides/advanced_tutorials/porting_cpus.mdx:50:7
    [...] See full stack trace in the browser, or rerun with --verbose.
```

Author: Thanks for the suggestions! I have resolved the build issues in the newest commit

  speed,
- enable pull-ups as appropriate for the platform,
- serialize access to the shared controller,
- generate START, RESTART, and STOP conditions correctly,
- translate RIOT flags into controller transactions,
- power the peripheral up in `i2c_acquire()` and down again in
  `i2c_release()` where practical.

### Generic high-level shape

```c
i2c_init(dev);

i2c_acquire(dev);
i2c_read_bytes(dev, addr, data, len, 0);
i2c_release(dev);
```

In the CPU driver, `i2c_init()`, `i2c_acquire()`, `i2c_read_bytes()`,
`i2c_write_bytes()`, and `i2c_release()` should implement correct START,
RESTART, ACK/NACK, and STOP handling for the underlying controller.

Correct handling of repeated starts, ACK/NACK conditions, and bus-busy recovery
is more important than adding optional optimizations early.

## General Recommendations

### Start small

Bring up the CPU with the smallest possible configuration first:

- boot to `cpu_init()`,
- confirm the vector table works,
- get `examples/basic/hello-world` running over UART,
- if UART output is missing, toggle an LED at key checkpoints in startup and
  `cpu_init()` to localize where execution stops,

> - validate one timer instance so `ztimer` works,

Reviewer 2: Which test would be advised to run?

configurations. This drastically reduces maintenance and lowers the risk of
subtle startup or interrupt bugs, as well as code divergation issues in the
future.

### Refer to the manual of the target CPU

Many CPUs are shipped with detailed manuals containing register-level
information about peripherals, clock trees, the boot sequence, interrupt
handling, and more. This information should be read, understood, and then mapped
to the RIOT framework.

### Keep board details out of the CPU driver

The CPU driver should not assume a specific LED pin, shield, or sensor layout.
Those belong in the board port. The CPU driver should consume generic
configuration from `periph_conf.h` and `periph_cpu.h`.

### Match the API exactly

Before implementing a peripheral, read the corresponding header in
`drivers/include/periph/`. The header documents not only the function
signatures, but also the expected semantics, error handling, and power
management model.

### Validate with basic applications

Useful first tests are:

- `examples/basic/hello-world` for startup and UART,
- `tests/periph_gpio` or GPIO-based samples for pin control,

> - timer-based examples for `periph_timer` and `ztimer`,

Reviewer 2: ```suggestion
- timer-based tests for `periph_timer` and `ztimer`,
  - refer to test applications in `tests/periph/timer` and `sys/ztimer*`
```

Something like this, to refer them to exact locations.

- a small SPI or I2C loopback / device test once the bus drivers are added.

## Further Reading

- RIOT guide: [board porting](/advanced_tutorials/porting_boards)
- RIOT peripheral APIs in `drivers/include/periph/`
- architecture-common code such as `cpu/cortexm_common/` when porting
  Cortex-M-based MCUs
- merged RIOT PRs adding CPU support, for reference and pattern matching:

>   <https://github.com/RIOT-OS/RIOT/pulls?q=is%3Apr+is%3Amerged+%22cpu%2F%22+%22boards%2F%22>

Reviewer: ```suggestion
  [see GitHub](https://github.com/RIOT-OS/RIOT/pulls?q=is%3Apr+is%3Amerged+%22cpu%2F%22+%22boards%2F%22)
```

To fix the build error:
```
11:07:49 [ERROR] Unexpected character `/` (U+002F) before local name, expected a character that can start a name, such as a letter, `$`, or `_` (note: to create a link in MDX, use `[text](url)`)
  Stack trace:
    at /home/cbuec/RIOTstuff/riot-vanillaice/RIOT/doc/guides/advanced_tutorials/porting_cpus.mdx:650:10
    [...] See full stack trace in the browser, or rerun with --verbose.
```
